backend/src/halo2_structs.rs

```diff
@@ -0,0 +1,99 @@
+use std::{
+    fs,
+    io::{self, Write},
+    path::Path,
+};
+
+use crate::{Backend, BackendBuilder, Proof};
+use ast::analyzed::Analyzed;
+use number::{DegreeType, FieldElement};
+
+pub struct Halo2Builder;
+impl<T: FieldElement> BackendBuilder<T> for Halo2Builder {
+    fn setup_from_params(&self, size: DegreeType) -> Box<dyn Backend<T>> {
+        Box::new(halo2::Halo2::new(size))
+    }
+
+    fn read(&self, mut input: &mut dyn io::Read) -> Result<Box<dyn Backend<T>>, io::Error> {
+        Ok(Box::new(halo2::Halo2::read(&mut input)?))
+    }
+}
+#[cfg(feature = "halo2")]
+impl<T: FieldElement> Backend<T> for halo2::Halo2<T> {
+    fn prove(
+        &self,
+        pil: &Analyzed<T>,
+        fixed: &[(&str, Vec<T>)],
+        witness: Option<&[(&str, Vec<T>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()> {
```

Comment:
Currently, the only thing the CLI does to the proof is write it to disk, and it is the backend who decides the file name. Also, that doesn't really work with Pilcom backend, as it doesn't really produce any proof, just writes some files.

Everything points to letting the backend handle the output...

Reply:
I think this is fine for this PR to not delay it further, but I also think ultimately the client should just receive the artifacts in memory and decide what to do with it.

Reply:
I disagree, if pilcom only writes files, it can write to a tempfile, read it and return the proof. The proof is super small, there is no reason to pass it around as a file path.